File: backend/app/schemas/level.py
"""
用户等级系统 Schema
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


# ==================== 行为枚举 ====================

class LevelAction(str, Enum):
    """规则引擎支持的用户行为类型. 新增行为必须同步更新 LEVEL_RULES."""
    POST_CREATED       = "post_created"
    COMMUNITY_FOLLOWED = "community_followed"
    POST_LIKED         = "post_liked"
    USER_FOLLOWED      = "user_followed"
    WANT_CLICKED       = "want_clicked"
    STORE_COMMENTED    = "store_commented"
    ARCHIVE_UPLOADED   = "archive_uploaded"


# ==================== 任务描述 (前端展示用) ====================

class LevelTaskSpec(BaseModel):
    """单个升级任务的目标描述"""
    action: str = Field(..., description="行为 key (对应 counters 里的字段)")
    target: int = Field(..., description="达标需要的累计次数")
    label:  str = Field(..., description="给用户看的任务文案")


class LevelSpec(BaseModel):
    """某个等级的完整定义"""
    level:     int
    title:     str
    subtitle:  str                = ""
    tasks:     List[LevelTaskSpec] = []
    benefit:   Optional[str]      = None
    # 'AUTO'   - 系统自动升级 (Lv1/2/3)
    # 'AUDIT'  - 达标后进入审核队列 (Lv4)
    # 'MANUAL' - 仅 admin 人工赋予 (Lv5)
    mode:      str                = "AUTO"


# ==================== 状态响应 ====================

class LevelTaskProgress(BaseModel):
    """单个任务的当前进度"""
    action:    str
    label:     str
    target:    int
    progress:  int
    completed: bool


class UserLevelStatus(BaseModel):
    """当前用户等级总览 (我的等级页面用)"""
    userId:           int
    currentLevel:     int
    pendingLevel:     Optional[int] = None
    lastLevelUpAt:    Optional[str] = None
    nextLevel:        Optional[int] = None
    nextLevelTitle:   Optional[str] = None
    nextLevelBenefit: Optional[str] = None
    # 距离 nextLevel 的任务进度;已完成或已到顶时为空数组
    nextTasks:        List[LevelTaskProgress] = []
    # 已解锁的权益列表
    benefits:         List["UserBenefitInfo"] = []


# ==================== 抽奖 ====================

class LotteryPrize(BaseModel):
    """奖品配置项 (lottery_rounds.prize_config 中的一项)"""
    prizeId: str
    name:    str
    quota:   int = Field(..., ge=0, description="总名额")
    meta:    Optional[Dict[str, Any]] = None


class LotteryRoundInfo(BaseModel):
    """抽奖期数信息"""
    id:            int
    month:         str     # 'YYYY-MM'
    status:        str     # OPEN / DRAWN / CLOSED
    prizeConfig:   List[LotteryPrize] = []
    drawnAt:       Optional[str]      = None
    totalEntries:  int                = 0
    totalWinners:  int                = 0


class LotteryEntryInfo(BaseModel):
    """用户在某一期的参与/中奖状态"""
    roundId:    int
    month:      str
    entered:    bool
    isWinner:   bool
    prizeId:    Optional[str]          = None
    prizeName:  Optional[str]          = None
    prizeMeta:  Optional[Dict[str, Any]] = None
    roundStatus: str                    = "OPEN"


class AdminDrawWinnerRequest(BaseModel):
    """Admin 开奖时手动指派的单条中奖信息"""
    userId:   int
    prizeId:  str


class AdminDrawLotteryRequest(BaseModel):
    """Admin 开奖请求:  winners 为 null 表示按 prize_config 随机抽"""
    winners: Optional[List[AdminDrawWinnerRequest]] = None


class AdminCreateRoundRequest(BaseModel):
    """Admin 建期 / 更新奖池请求"""
    month:       Optional[str] = Field(None, description="'YYYY-MM',不传则取当月")
    prizeConfig: List[LotteryPrize] = []


# ==================== 权益 ====================

class UserBenefitInfo(BaseModel):
    """用户持有的权益"""
    benefitId:    int
    benefitType:  str
    name:         str
    description:  str = ""
    quota:        int
    used:         int
    remaining:    int


class RedeemTicketRequest(BaseModel):
    """核销免费门票"""
    objectType: str              = "EVENT"
    objectId:   Optional[str]    = None
    meta:       Optional[Dict[str, Any]] = None


class RedeemTicketResponse(BaseModel):
    redemptionId: int
    remaining:    int


# ==================== 升级审批 ====================

class UpgradeRequestInfo(BaseModel):
    id:          int
    userId:      int
    username:    Optional[str] = None
    targetLevel: int
    status:      str
    remark:      str = ""
    createdAt:   str
    reviewedAt:  Optional[str] = None


class AdminReviewUpgradeRequest(BaseModel):
    approve: bool
    remark:  str = ""


class AdminGrantLevelRequest(BaseModel):
    """Admin 直接赋予等级 (用于 Lv5)"""
    level:  int = Field(..., ge=1, le=5)
    remark: str = ""


class AdminBackfillRequest(BaseModel):
    """存量用户等级回填请求.

    - user_id=None + limit=None  -> 全量扫描
    - user_id=X                  -> 单用户回填
    - limit=N                    -> 最多处理 N 人 (用于分批压测)
    """
    userId: Optional[int] = None
    dryRun: bool          = False
    limit:  Optional[int] = None
    offset: int           = 0


# 允许 forward reference
UserLevelStatus.model_rebuild()

File: web/src/lib/services/level.ts
/**
 * Web 端用户等级 / 月度抽奖 / 权益 API 客户端.
 *
 * 对齐 frontend/src/services/levelService.ts 的类型与方法语义,
 * 确保 iOS / Android / Web 走同一套后端契约.
 *
 * 使用场景:
 *   - `levelApi.*`         → 用户侧 (我的等级页, 主页徽章, 月度抽奖入口)
 *   - `adminLevelApi.*`    → 运营侧 (Lv4 审批, Lv5 授予, 月度开奖)
 */

import { apiClient } from "../api-client";

// ==================== 类型 ====================

export interface LevelTaskSpec {
  action: string;
  target: number;
  label: string;
}

export interface LevelSpec {
  level: number;
  title: string;
  subtitle: string;
  tasks: LevelTaskSpec[];
  benefit: string | null;
  mode: "AUTO" | "AUDIT" | "MANUAL";
}

export interface LevelTaskProgress {
  action: string;
  label: string;
  target: number;
  progress: number;
  completed: boolean;
}

export interface UserBenefitInfo {
  benefitId: number;
  benefitType: string;
  name: string;
  description: string;
  quota: number;
  used: number;
  remaining: number;
}

export interface UserLevelStatus {
  userId: number;
  currentLevel: number;
  pendingLevel: number | null;
  lastLevelUpAt: string | null;
  nextLevel: number | null;
  nextLevelTitle: string | null;
  nextLevelBenefit: string | null;
  nextTasks: LevelTaskProgress[];
  benefits: UserBenefitInfo[];
}

export interface LotteryPrize {
  prizeId: string;
  name: string;
  quota: number;
  meta?: Record<string, unknown> | null;
}

export interface LotteryRoundInfo {
  id: number;
  month: string;
  status: "OPEN" | "DRAWN" | "CLOSED";
  prizeConfig: LotteryPrize[];
  drawnAt: string | null;
  totalEntries: number;
  totalWinners: number;
}

export interface LotteryEntryInfo {
  roundId: number;
  month: string;
  entered: boolean;
  isWinner: boolean;
  prizeId: string | null;
  prizeName: string | null;
  prizeMeta: Record<string, unknown> | null;
  roundStatus: "OPEN" | "DRAWN" | "CLOSED";
}

export interface CurrentLotteryPayload {
  /**
   * Admin 全站开关. true = 抽奖功能开启;
   * false / 缺省 = 已被关闭 (后端默认), `round` / `entry` 为 null,
   * 客户端必须隐藏所有抽奖入口与卡片.
   */
  enabled?: boolean;
  round: LotteryRoundInfo | null;
  entry: LotteryEntryInfo | null;
}

export interface UpgradeRequestInfo {
  id: number;
  userId: number;
  username: string | null;
  targetLevel: number;
  status: "PENDING" | "APPROVED" | "REJECTED";
  remark: string;
  createdAt: string;
  reviewedAt: string | null;
}

export interface RedeemTicketResponse {
  redemptionId: number;
  remaining: number;
}

export interface BackfillUserResult {
  userId: number;
  beforeLevel: number;
  afterLevel: number;
  pendingLevel: number | null;
  counters: Record<string, number>;
  dryRun: boolean;
}

export interface BackfillSummary {
  scanned: number;
  upgraded: number;
  pendingCreated: number;
  errors: number;
  levelDistribution: Record<string, number>;
  dryRun: boolean;
}

export type BackfillResponse =
  | { scope: "single"; user: BackfillUserResult }
  | { scope: "all"; summary: BackfillSummary };

export interface LevelUserRow {
  userId: number;
  username: string;
  avatarUrl: string;
  currentLevel: number;
  pendingLevel: number | null;
  lastLevelUpAt: string | null;
  merchant?: { storeId: number; status: string } | null;
}

export interface ListLevelUsersResponse {
  users: LevelUserRow[];
  total: number;
  page: number;
  pageSize: number;
}

// ==================== 用户侧 ====================

export const levelApi = {
  getRules: () => apiClient.get<LevelSpec[]>("/api/levels/rules"),

  getMyLevel: () => apiClient.get<UserLevelStatus>("/api/levels/me"),

  getUserLevel: (userId: number | string) =>
    apiClient.get<{ userId: number; currentLevel: number }>(
      `/api/levels/users/${userId}/summary`,
    ),

  getCurrentLottery: () =>
    apiClient.get<CurrentLotteryPayload>("/api/lottery/current"),

  getLotteryHistory: (limit = 12) =>
    apiClient.get<LotteryRoundInfo[]>(`/api/lottery/history`, { limit }),

  getMyBenefits: () => apiClient.get<UserBenefitInfo[]>("/api/benefits/me"),

  /**
   * Lv4 免费门票核销.
   *
   * **红线提醒**: 本方法**只能在事件报名页**(未来实装)被调用,
   * "我的等级"页严禁暴露核销按钮. 这一约束在 iOS 端同样生效,
   * 这里留此方法是为了未来 Web 事件页接入时走同一契约.
   */
  redeemFreeTicket: (payload: {
    objectType?: string;
    objectId?: string;
    meta?: Record<string, unknown>;
  }) =>
    apiClient.post<RedeemTicketResponse>(
      "/api/benefits/free-ticket/redeem",
      {
        objectType: payload.objectType ?? "EVENT",
        objectId: payload.objectId,
        meta: payload.meta,
      },
    ),
};

// ==================== Admin 侧 ====================

export const adminLevelApi = {
  listUpgradeRequests: () =>
    apiClient.get<UpgradeRequestInfo[]>("/api/admin/levels/upgrade-requests"),

  reviewUpgradeRequest: (
    requestId: number,
    approve: boolean,
    remark = "",
  ) =>
    apiClient.post<null>(
      `/api/admin/levels/upgrade-requests/${requestId}/review`,
      { approve, remark },
    ),

  grantLevel: (userId: number, level: number, remark = "") =>
    apiClient.post<null>(`/api/admin/levels/users/${userId}/grant`, {
      level,
      remark,
    }),

  listRounds: (limit = 24) =>
    apiClient.get<LotteryRoundInfo[]>("/api/admin/lottery/rounds", { limit }),

  upsertRound: (month: string | null, prizeConfig: LotteryPrize[]) =>
    apiClient.post<LotteryRoundInfo>("/api/admin/lottery/rounds", {
      month,
      prizeConfig,
    }),

  syncEntries: (roundId: number) =>
    apiClient.post<{ added: number }>(
      `/api/admin/lottery/rounds/${roundId}/sync-entries`,
    ),

  drawRound: (
    roundId: number,
    winners: Array<{ userId: number; prizeId: string }> | null = null,
  ) =>
    apiClient.post<{ winners: number }>(
      `/api/admin/lottery/rounds/${roundId}/draw`,
      { winners },
    ),

  /**
   * 存量用户等级回填 (幂等).
   *
   * - 不传 userId / limit -> 全量扫描
   * - 传 userId           -> 仅该用户
   * - dryRun=true         -> 只计算不写库
   */
  backfillLevels: (payload: {
    userId?: number;
    dryRun?: boolean;
    limit?: number;
    offset?: number;
  } = {}) =>
    apiClient.post<BackfillResponse>("/api/admin/levels/backfill", {
      userId: payload.userId ?? null,
      dryRun: payload.dryRun ?? false,
      limit: payload.limit ?? null,
      offset: payload.offset ?? 0,
    }),

  /**
   * 分页拉所有用户等级 (current_level 降序).
   *
   * @param level  `null` = 全部, `0` = 仅 Lv0 (未达 Lv1), `1..5` = 精确等级
   */
  listUsersByLevel: (payload: {
    page?: number;
    pageSize?: number;
    level?: number | null;
  } = {}) => {
    const query: Record<string, string | number> = {
      page: payload.page ?? 1,
      pageSize: payload.pageSize ?? 20,
    };
    if (payload.level !== undefined && payload.level !== null) {
      query.level = payload.level;
    }
    return apiClient.get<ListLevelUsersResponse>(
      "/api/admin/levels/users",
      query,
    );
  },
};
